runtime/compiler/ilgen/Walker.cpp

```diff
@@ -2369,13 +2371,26 @@ TR_J9ByteCodeIlGenerator::genCheckCast(int32_t cpIndex)
    {
    if (TR::Compiler->om.areValueTypesEnabled() && TR::Compiler->cls.isClassRefValueType(comp(), method()->classOfMethod(), cpIndex))
       {
+      void * classObject = method()->getClassFromConstantPool(comp(), cpIndex);
```

Comment:
Why can we not use loadClassObject here?

Reply:
`loadClassObject()` pushes a node to the stack. The symbolic reference would be pushed on the top of the stack instead of the `objectref` before the `NULLCHK` node creation. `passThruNode’`s child becomes the symbolic reference, instead of `objectRef`. `NULLCHK`’s grandchild node ends up being wrong.

Reply:
@a7ehuo thanks for the explanation - rather than duplicating this logic inline, I think we should refactor loadSymbol into a loadSymbol and loadSymbolNoPush or something like that. loadSymbol can then call loadSymbolNoPush and then push. That way we can keep all that logic together and avoid creating a case to be missed if we have to change anything about how this works...

Reply:
I looked at the code again and I wonder if the followings achieve the same thing without duplicating the logic or refactoring `loadSymbol`? 

```c++
   if (TR::Compiler->om.areValueTypesEnabled() && TR::Compiler->cls.isClassRefValueType(comp(), method()->classOfMethod(), cpIndex))
      {
      TR::Node * objNode = _stack->top();

      loadClassObject(cpIndex);

      TR::Node *passThruNode = TR::Node::create(TR::PassThrough, 1, objNode);
      genTreeTop(genNullCheck(passThruNode));
      }
```

Reply:
so the issue will be that the push in loadClassObject will be onto the real operand stack and that is likely what you want to avoid - you could potentially pop it, but why push to pop... So I think the stack will still be the wrong shape after this if the goal is to load the SymRef without pushing its value onto the simulated operand stack.

Reply:
On second thought you may be right - test the sequence, but your proposal would be preferable if it works.

Reply:
The new sequence https://github.com/eclipse/openj9/pull/10068#discussion_r453832652 looks good to me. The operation on the stack as below is the same as duplicating `loadClassObject` logic inline: 

```
Enter genCheckCast(int32_t cpIndex): stack top aconst
===>loadClassObject(): stack top aconst
              - push loadaddr
===>Enter genCheckCast(): stack top loadaddr
               Enter genNodeAndPopChildren(): stack top loadaddr
	            - pop loadaddr
	            - pop aconst
               Exit genNodeAndPopChildren(): stack empty 
	       push aconst
===>Exit genCheckCast(): stack top aconst
Exit genCheckCast(int32_t cpIndex): stack top aconst
```

```
  ============================================================
   1: JBcheckcast
   /--- trees inserted ------------------------
      n5n      (  0)  ResolveCHK [#307]                                                                    [0x7f13f1057540] bci=[-1,1,-] rc=0 vc=0 vn=- li=- udi=- nc=1
      n4n      (  2)    loadaddr  unknown class object[#366  unresolved Static] [flags 0x18307 0x0 ]       [0x7f13f10574f0] bci=[-1,1,-] rc=2 vc=0 vn=- li=- udi=- nc=0
      n7n      (  0)  NULLCHK on n3n [#32]                                                                 [0x7f13f10575e0] bci=[-1,1,-] rc=0 vc=0 vn=- li=- udi=- nc=1
      n6n      (  1)    PassThrough                                                                        [0x7f13f1057590] bci=[-1,1,-] rc=1 vc=0 vn=- li=- udi=- nc=1
      n3n      (  2)      aconst NULL                                                                      [0x7f13f10574a0] bci=[-1,0,-] rc=2 vc=0 vn=- li=- udi=- nc=0
      n8n      (  0)  checkcast [#86]                                                                      [0x7f13f1057630] bci=[-1,1,-] rc=0 vc=0 vn=- li=- udi=- nc=2
      n3n      (  2)    ==>aconst NULL
      n4n      (  2)    ==>loadaddr
   /--- stack after ------------------------
   @0 n3n      (  2)  ==>aconst NULL
  ============================================================

```